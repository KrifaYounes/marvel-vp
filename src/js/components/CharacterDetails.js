import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { loadCharacters } from '../actions/characters-actions';
import { Grid, Row, Col } from 'react-flexbox-grid';


class CharacterDetails extends Component {
    componentDidMount() {
        const characterId = this.props.match.url;
        this.props.loadCharacters(characterId);
    }

    render() {
        const { characters, isLoading } = this.props;
        const character = characters[0];

        if (isLoading) {
            return ( <div> Loading </div>);
        } else if (!character) {
            return ( <div> Heroe Not Found </div>);

        } else {
            const { name, description, thumbnail, series, comics} = character;
            const imgSource = thumbnail.path + '.' + thumbnail.extension ;

            const serieItems = series.items.map( (serie, index) => {
                    return ( <li key={index}>{serie.name}</li> )
                });

            const comicItems = comics.items.map( (comic, index) => {
                    return ( <li key={index}>{comic.name}</li> )
            });

            return (
                <div>
                    <Grid>
                        <Row>
                            <Col>
                                <div className='character'>
                                    <img src={imgSource} alt="MARVEL"/>
                                </div>
                            </Col>

                            <Col xs={8}>
                                <div className="character-description">
                                    <div>
                                        {name}
                                    </div> <br/>
                                    <div>
                                        {description}
                                    </div>
                                </div>
                            </Col>
                        </Row>

                        <Row>
                            <Column title ="Series" items={serieItems} />
                            <Column title ="Comics" items={comicItems} />
                        </Row>
                    </Grid>

                </div>
            );
        }
    }

}


const Column = ({title, items}) => (
    <Col xs={6}>
        <div className="box-shadow">
            <h3> {title} </h3>
            <ul> {items.length > 0 ? items : 'Not available'} </ul>
        </div>
    </Col>
);


CharacterDetails.CharacterDetails = {
    characters: PropTypes.object,
    isLoading: PropTypes.bool
};

function mapStateToProps(state) {
    const characters = state.charactersReducer.characters;
    const isLoading = state.charactersReducer.isLoading;

    return {
        characters,
        isLoading
    };
}

export default connect(mapStateToProps, { loadCharacters } )(CharacterDetails);

