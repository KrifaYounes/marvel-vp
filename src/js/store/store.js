import { createStore, applyMiddleware } from 'redux'
import { syncHistoryWithStore } from 'react-router-redux';
import { createHashHistory } from 'history';
import reducers from '../reducers/reducers';
import thunkMiddleware from 'redux-thunk';

export const history = createHashHistory();

const store = createStore(reducers, applyMiddleware(thunkMiddleware));

syncHistoryWithStore(history, store);

export default store;
