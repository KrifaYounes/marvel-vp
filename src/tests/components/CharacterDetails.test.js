import React from 'react';
import CharacterDetails from '../../js/components/CharacterDetails';
import { shallow } from 'enzyme';
import enzymeToJSON from 'enzyme-to-json';
import store from '../../js/store/store';

test('CharacterDetails Component Snapshot', () => {
    const character = {
        urls : [],
        thumbnail : {
            path : 'http://marvel.com/image',
            extension: 'jpeg'
        },
        descriptoin: 'Marvel Heroe Description',
        name: 'Marvel Heroe Name'
    };

    const wrapper = shallow(<CharacterDetails character={character} isLoading={false} store={store}/>);
    //must match the snapshot

    expect(enzymeToJSON(wrapper)).toMatchSnapshot();
});


test('CharacterDetails Component Snapshot without character', () => {
    const character = null;

    const wrapper = shallow(<CharacterDetails character={character} isLoading={false} store={store}/>);
    //must match the snapshot

    expect(enzymeToJSON(wrapper)).toMatchSnapshot();
});


test('CharacterDetails Component Snapshot Loading', () => {
    const character = null;

    const wrapper = shallow(<CharacterDetails character={character} isLoading={true} store={store}/>);
    //must match the snapshot

    expect(enzymeToJSON(wrapper)).toMatchSnapshot();
});
