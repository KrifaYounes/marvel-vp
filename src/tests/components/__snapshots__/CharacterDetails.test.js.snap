// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetails Component Snapshot 1`] = `
<CharacterDetails
  character={
    Object {
      "descriptoin": "Marvel Heroe Description",
      "name": "Marvel Heroe Name",
      "thumbnail": Object {
        "extension": "jpeg",
        "path": "http://marvel.com/image",
      },
      "urls": Array [],
    }
  }
  characters={Array []}
  isLoading={false}
  loadCharacters={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`CharacterDetails Component Snapshot Loading 1`] = `
<CharacterDetails
  character={null}
  characters={Array []}
  isLoading={false}
  loadCharacters={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`CharacterDetails Component Snapshot without character 1`] = `
<CharacterDetails
  character={null}
  characters={Array []}
  isLoading={false}
  loadCharacters={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
